sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -618,8 +619,84 @@ class DataFrame protected[sql](
    */
   @Experimental
   override def saveAsTable(tableName: String): Unit = {
-    sqlContext.executePlan(
-      CreateTableAsSelect(None, tableName, logicalPlan, allowExisting = false)).toRdd
+    val dataSourceName = sqlContext.conf.defaultDataSourceName
+    saveAsTable(tableName, dataSourceName, ("path", sqlContext.defaultTableFilePath(tableName)))
+  }
+
+  /**
+   * :: Experimental ::
+   * Creates a table from the the contents of this DataFrame based on a given data source and
+   * a set of options. This will fail if the table already exists.
+   *
+   * Note that this currently only works with DataFrames that are created from a HiveContext as
+   * there is no notion of a persisted catalog in a standard SQL context.  Instead you can write
+   * an RDD out to a parquet file, and then register that file as a table.  This "table" can then
+   * be the target of an `insertInto`.
+   */
+  @Experimental
+  override def saveAsTable(
+      tableName: String,
+      dataSourceName: String,
+      options: (String, String)*): Unit = {
+    val cmd =
+      CreateTableUsingAsLogicalPlan(
+        tableName,
+        dataSourceName,
+        temporary = false,
+        options.toMap,
+        allowExisting = false,
+        logicalPlan)
+
+    sqlContext.executePlan(cmd).toRdd
+  }
+
+  /**
+   * :: Experimental ::
+   * Creates a table from the the contents of this DataFrame based on a given data source and
+   * a set of options. This will fail if the table already exists.
+   *
+   * Note that this currently only works with DataFrames that are created from a HiveContext as
+   * there is no notion of a persisted catalog in a standard SQL context.  Instead you can write
+   * an RDD out to a parquet file, and then register that file as a table.  This "table" can then
+   * be the target of an `insertInto`.
+   */
+  @Experimental
+  override def saveAsTable(
+      tableName: String,
+      dataSourceName: String,
+      options: java.util.Map[String, String]): Unit = {
+    saveAsTable(tableName, dataSourceName, options.toSeq:_*)
+  }
+
+  @Experimental
+  override def save(path: String): Unit = {
+    save(path, false)
+  }
+
+  @Experimental
+  override def save(path: String, overwrite: Boolean): Unit = {
+    val dataSourceName = sqlContext.conf.defaultDataSourceName
+    save(dataSourceName, overwrite, ("path" -> path))
+  }
+
+  @Experimental
+  override def save(
+      dataSourceName: String,
+      overwrite: Boolean,
+      options: (String, String)*): Unit = {
```

Comment:
here too. let's disambiguate it by having option: (String, String) followed by options: (String, String)